kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/check/consistency/algorithm/CRC32MatchDataConsistencyCalculateAlgorithm.java

```diff
@@ -37,51 +40,137 @@
+            beginId = getBeginIdFromUniqueKey(uniqueKey.getDataType());
+        } else {
+            beginId = previousCalculatedResult.getMaxUniqueKeyValue();
+        }
+        Object endId = getMaxUniqueKeyValue(sqlBuilder, parameter);
+        if (null == endId) {
+            return Optional.empty();
+        }
+        List<CalculatedItem> calculatedItems = parameter.getColumnNames().stream().map(each -> calculateCRC32(sqlBuilder, parameter, each, beginId, endId)).collect(Collectors.toList());
+        int recordsCount = calculatedItems.get(0).getRecordsCount();
+        return Optional.of(new CalculatedResult(endId, recordsCount, calculatedItems.stream().map(CalculatedItem::getCrc32).collect(Collectors.toList())));
+    }
+    
+    private Object getBeginIdFromUniqueKey(final int columnType) {
+        if (PipelineJdbcUtils.isStringColumn(columnType)) {
+            return "!";
+        } else {
+            return Integer.MIN_VALUE;
+        }
     }
     
-    private CalculatedItem calculateCRC32(final PipelineSQLBuilder sqlBuilder, final DataConsistencyCalculateParameter parameter, final String columnName) {
+    private Object getMaxUniqueKeyValue(final PipelineSQLBuilder sqlBuilder, final DataConsistencyCalculateParameter parameter) {
+        String schemaName = parameter.getSchemaName();
+        String logicTableName = parameter.getLogicTableName();
+        String cacheKeyPrefix = "uniqueKey-" + (null == parameter.getPreviousCalculatedResult() ? "first" : "later") + "-";
+        String cacheKey = cacheKeyPrefix + parameter.getDatabaseType() + "-" + (null != schemaName && DatabaseTypeFactory.getInstance(parameter.getDatabaseType()).isSchemaAvailable()
+                ? schemaName + "." + logicTableName
+                : logicTableName);
+        String sql = sqlCache.computeIfAbsent(cacheKey, s -> sqlBuilder.buildChunkedQueryUniqueKeySQL(schemaName, logicTableName, parameter.getUniqueKey().getName(),
+                null == parameter.getPreviousCalculatedResult()));
+        CalculatedResult previousCalculatedResult = (CalculatedResult) parameter.getPreviousCalculatedResult();
+        try (
+                Connection connection = parameter.getDataSource().getConnection();
+                PreparedStatement preparedStatement = setCurrentStatement(connection.prepareStatement(sql))) {
+            preparedStatement.setFetchSize(chunkSize);
+            if (null == previousCalculatedResult) {
+                preparedStatement.setInt(1, chunkSize);
+            } else {
+                preparedStatement.setObject(1, previousCalculatedResult.getMaxUniqueKeyValue());
+                preparedStatement.setInt(2, chunkSize);
+            }
+            Object maxUniqueKeyValue = null;
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                while (resultSet.next()) {
+                    maxUniqueKeyValue = resultSet.getObject(1);
+                }
+            }
```

Comment:
It's not efficient enough to get maximum value of unique key. Could we query `max(uniqueKey)` in SQL?

Reply:
relation issue #21540, maybe we can improve later?

Reply:
It might hurt performance too much, if check crc32 block by block just benefit for check progress, it might not worth to do it.

So we need try to find another way, or else just give up it for now.